File: 프로그래머스/unrated/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> answer;
    int n = sequence.size();
    int st = 0, ed = 0;
    int sum = 0;
    int stIdx = 0, edIdx = 0;
    int length = 1000000001;
    while(st < n && ed <= n){
        if(sum >= k){
            sum -= sequence[st++];
        }
        else sum += sequence[ed++];
        if(sum == k) {
            int tempLength = ed-st;
            if(tempLength < length){
                length = tempLength;
                stIdx = st;
                edIdx = ed-1;
            }
            else if(tempLength == length){
                if(stIdx > st){
                    stIdx = st;
                    edIdx = ed-1;
                }
            }
        }
    }
    answer.push_back(stIdx);
    answer.push_back(edIdx);
    return answer;
}

File: 프로그래머스/unrated/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.java

class Solution {
    public int[] solution(int[] sequence, int k) {
        int[] answer = new int[]{0, 0};
        int ed = 0;
        int len = Integer.MAX_VALUE;
        int n = sequence.length;
        int sum = sequence[0];

        for (int st = 0; st < n; st++) {
            while (ed < n && sum < k) {
                ed++;
                if (ed != n) sum += sequence[ed];
            }
            if (ed == n) break;
            if (sum == k) {
                if (ed - st < len) {
                    len = ed - st;
                    answer[0] = st;
                    answer[1] = ed;
                } else if ((ed - st == len) && st < answer[0]) {
                    answer[0] = st;
                    answer[1] = ed;
                }
            }
            sum -= sequence[st];
        }
        return answer;
    }
}